pkg/utils/service.go

```diff
@@ -0,0 +1,60 @@
+package utils
+
+import (
+	"fmt"
+	"strings"
+
+	v1core "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func ServiceForEndpoints(ci *cache.Indexer, ep *v1core.Endpoints) (interface{}, error) {
+	key, err := cache.MetaNamespaceKeyFunc(ep)
+	if err != nil {
+		return nil, err
+	}
+
+	item, exists, err := (*ci).GetByKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, fmt.Errorf("service resource doesn't exist for endpoints: %q", ep.Name)
+	}
+
+	return item, nil
+}
+
+// ServiceIsHeadless decides whether or not the this service is a headless service which is often useful to kube-router
+// as there is no need to execute logic on most headless changes. Function takes a generic interface as its input
+// parameter so that it can be used more easily in early processing if needed. If a non-service object is given,
+// function will return false.
+func ServiceIsHeadless(obj interface{}) bool {
+	if svc, _ := obj.(*v1core.Service); svc != nil {
+		if svc.Spec.Type == v1core.ServiceTypeClusterIP {
+			if ClusterIPIsNone(svc.Spec.ClusterIP) && containsOnlyNone(svc.Spec.ClusterIPs) {
```

Comment:
Is this enough? Dont we need `ClusterIPIsNoneOrBlank`? As per https://kubernetes.io/docs/reference/kubernetes-api/services-resources/service-v1/#ServiceSpec empty string is vali value as well.

Reply:
I'm definitely open to changing it, but according to the API, it seems that they hint that only "None" that indicates a headless service:
```
Setting this to "None" makes a "headless service" (no virtual IP), which is useful when direct endpoint connections are preferred and proxying is not required.
```

Again, I wish that the upstream docs were a little bit more definitive. As they don't really describe what setting it to `""` is supposed to do. Practically speaking, usually this indicates that the kube-controller-manager should allocate an IP at random.

Reply:
Yes, its confusing. Given its explicitly states set to None to make it headless, it make sense to stick to what you code.

I will rerun the CI and proceed for merge.